src/resolve/resolved-etc-hosts.c

```diff
@@ -431,7 +431,7 @@ int manager_etc_hosts_lookup(Manager *m, DnsQuestion* q, DnsAnswer **answer) {
         for (i = 0; i < bn->n_items; i++) {
                 _cleanup_(dns_resource_record_unrefp) DnsResourceRecord *rr = NULL;
 
-                if ((found_a && bn->items[i]->family != AF_INET) &&
+                if ((found_a && bn->items[i]->family != AF_INET) ||
                     (found_aaaa && bn->items[i]->family != AF_INET6))
                         continue;
```

Comment:
hmm, the old check was wrong, but the new one, too ;-)

`found_a` is true if the question supplied by the user included an "A" query, and `found_aaaa` if it included an "AAAA" query.

Now, we should skip over an answer record if:

 1. it is an A answer, but we had no A query

as well as if:

 2. it is an AAAA answer, but we had no AAAA query.

The correct check hence needs to be:

```c
(bn->items[i]->family == AF_INET && !found_a) ||
(bn->items[i]->family == AF_INET6 && !found_aaaa)
```

Following the cumulative law and De Morgan's law this could also be written as:

```c
(found_a || b->items[i]->family != AF_INET) &&
(found_aaaa || b->items[i]->family != AF_INET6)
```

But neither the way the original code put it, nor in the version of the patch...

Does this make sense?